File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    List<List<String>> res = new ArrayList<>();
    List<String> list = new ArrayList<>();
    public List<List<String>> partition(String s) {
        backtracking(s, 0);
        return res;
    }

    private void backtracking(String s, int start) {
        if (start == s.length()) {
            res.add(new ArrayList<String>(list));
            return;
        }

        for (int i = start; i < s.length(); i++) {
            if (isPalindrome(s, start, i)) {
                list.add(s.substring(start, i + 1));
                backtracking(s, i + 1);
                list.remove(list.size() - 1);

            }

            
            
        }
    }

    private boolean isPalindrome(String s, int start, int end) {
       while (start < end) {
           if (s.charAt(start) != s.charAt(end)) {
               return false;
           }
           start++;
           end--;
       }

       return true;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    const res = [];
    const substrings = [];
    
    function backtrack(start) {
        if (start === s.length) {
            res.push([...substrings]);
        }
        
        for (let i = start; i < s.length; i++) {
            if (isValid(start, i)) {
                substrings.push(s.substring(start, i + 1));
                backtrack(i + 1);
                substrings.pop();
                
            }
        }
    }
    
    function isValid(start, end) {
        if (end < start) {
            return false;
        } else if (end == start) {
            return true;
        } else {
            while (start < end) {
                if (s[start] !== s[end]) {
                    return false;
                }
                start++;
                end--;
            }
            
            return true;
        }
    }
    
    backtrack(0);
    return res;
};